src/target/xtensa/xtensa.c

```diff
@@ -975,58 +1059,73 @@ int xtensa_fetch_all_regs(struct target *target)
-		xtensa_queue_exec_ins(xtensa, XT_INS_WSR(XT_SR_DDR, XT_REG_A3));
-		xtensa_queue_dbg_reg_read(xtensa, NARADR_DDR, regvals[XT_REG_IDX_CPENABLE]);
+		xtensa_queue_exec_ins(xtensa, XT_INS_RSR(xtensa, xtensa_regs[XT_REG_IDX_CPENABLE].reg_num, XT_REG_A3));
+		xtensa_queue_exec_ins(xtensa, XT_INS_WSR(xtensa, XT_SR_DDR, XT_REG_A3));
+		xtensa_queue_dbg_reg_read(xtensa, NARADR_DDR, (uint8_t *)&regvals[XT_REG_IDX_CPENABLE]);
 	}
-	int res = jtag_execute_queue();
+	res = jtag_execute_queue();
 	if (res != ERROR_OK) {
 		LOG_ERROR("Failed to read ARs (%d)!", res);
 		return res;
 	}
 	xtensa_core_status_check(target);
 
+	a3 = buf_get_u32(a3_buf, 0, 32);
+
 	if (xtensa->core_config->coproc)
-		cpenable = buf_get_u32(regvals[XT_REG_IDX_CPENABLE], 0, 32);
+		cpenable = buf_get_u32((uint8_t *)&regvals[XT_REG_IDX_CPENABLE], 0, 32);
 	/* We're now free to use any of A0-A15 as scratch registers
 	 * Grab the SFRs and user registers first. We use A3 as a scratch register. */
-	for (unsigned int i = 0; i < XT_NUM_REGS; i++) {
-		if (xtensa_reg_is_readable(xtensa_regs[i].flags, cpenable) && reg_list[i].exist &&
-			(xtensa_regs[i].type == XT_REG_SPECIAL ||
-				xtensa_regs[i].type == XT_REG_USER || xtensa_regs[i].type == XT_REG_FR)) {
-			if (xtensa_regs[i].type == XT_REG_USER) {
-				xtensa_queue_exec_ins(xtensa, XT_INS_RUR(xtensa_regs[i].reg_num, XT_REG_A3));
-			} else if (xtensa_regs[i].type == XT_REG_FR) {
-				xtensa_queue_exec_ins(xtensa, XT_INS_RFR(xtensa_regs[i].reg_num, XT_REG_A3));
-			} else {	/*SFR */
-				unsigned int reg_num = xtensa_regs[i].reg_num;
-				if (reg_num == XT_PC_REG_NUM_BASE) {
+	for (unsigned int i = 0; i < reg_list_size; i++) {
+		struct xtensa_reg_desc *rlist = (i < XT_NUM_REGS) ? xtensa_regs : xtensa_optregs;
+		uint32_t ridx = (i < XT_NUM_REGS) ? i : i - XT_NUM_REGS;
+		if (xtensa_reg_is_readable(rlist[ridx].flags, cpenable) && rlist[ridx].exist) {
+			bool reg_fetched = true;
+			unsigned int reg_num = rlist[ridx].reg_num;
+			switch (rlist[ridx].type) {
+			case XT_REG_USER:
+				xtensa_queue_exec_ins(xtensa, XT_INS_RUR(xtensa, reg_num, XT_REG_A3));
+				break;
+			case XT_REG_FR:
+				xtensa_queue_exec_ins(xtensa, XT_INS_RFR(xtensa, reg_num, XT_REG_A3));
+				break;
+			case XT_REG_SPECIAL:
+				if (reg_num == XT_PC_REG_NUM_VIRTUAL) {
 					/* reg number of PC for debug interrupt depends on NDEBUGLEVEL */
-					reg_num += xtensa->core_config->debug.irq_level;
+					reg_num = (XT_PC_REG_NUM_BASE + xtensa->core_config->debug.irq_level);
 				}
-				xtensa_queue_exec_ins(xtensa, XT_INS_RSR(reg_num, XT_REG_A3));
+				xtensa_queue_exec_ins(xtensa, XT_INS_RSR(xtensa, reg_num, XT_REG_A3));
+				break;
+			default:
+				reg_fetched = false;
+			}
+			if (reg_fetched) {
+				xtensa_queue_exec_ins(xtensa, XT_INS_WSR(xtensa, XT_SR_DDR, XT_REG_A3));
+				xtensa_queue_dbg_reg_read(xtensa, NARADR_DDR, (uint8_t *)&regvals[i]);
+				if (debug_dsrs)
+					xtensa_queue_dbg_reg_read(xtensa, NARADR_DSR, (uint8_t *)&dsrs[i]);
```

Comment:
host endiannes here, above and below in more lines

Reply:
Please see above comment.